File: backend/routers/servers.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload

from ..core.database import get_db
from ..models.server import (
    Server,
    Channel,
    ChannelType,
    ChannelGroup,
    Message,
    Attachment,
)
from .deps import CurrentUser, AdminUser
from .schemas import UTCDateTimeModel
from ..core.permissions import check_server_access, check_channel_visibility
from ..utils import extract_mentioned_usernames


router = APIRouter(prefix="/api/servers", tags=["servers"])


class ServerCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    icon: str | None = None


class ServerUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=100)
    icon: str | None = None
    min_server_level: int | None = None  # 1-4
    min_internal_level: int | None = None  # 1-2


class ServerResponse(BaseModel):
    id: int
    name: str
    icon: str | None
    owner_id: int
    min_server_level: int = 1
    min_internal_level: int = 1

    class Config:
        from_attributes = True


class ChannelResponse(BaseModel):
    id: int
    name: str
    type: str
    position: int
    top_position: int
    group_id: int | None = None
    visibility_min_server_level: int = 1
    visibility_min_internal_level: int = 1
    speak_min_server_level: int = 1
    speak_min_internal_level: int = 1

    class Config:
        from_attributes = True


class ServerDetailResponse(ServerResponse):
    channels: list[ChannelResponse]
    min_server_level: int = 1
    min_internal_level: int = 1


@router.get("", response_model=list[ServerResponse])
async def list_servers(user: CurrentUser, db: AsyncSession = Depends(get_db)):
    """List all servers user has access to based on SSO permission levels."""
    result = await db.execute(select(Server).order_by(Server.id))
    servers = result.scalars().all()
    
    # Filter servers based on user's permission levels
    filtered_servers = [
        server for server in servers
        if check_server_access(
            user,
            server.min_server_level,
            server.min_internal_level
        )
    ]
    
    return filtered_servers


@router.post("", response_model=ServerResponse, status_code=status.HTTP_201_CREATED)
async def create_server(
    payload: ServerCreate, user: AdminUser, db: AsyncSession = Depends(get_db)
):
    """Create a new server (admin only)."""
    server = Server(name=payload.name, icon=payload.icon, owner_id=user["id"])
    db.add(server)
    await db.flush()

    # Create default channels (ungrouped, so use top_position)
    general_text = Channel(
        server_id=server.id,
        name="general",
        type=ChannelType.TEXT,
        position=0,
        top_position=0,
    )
    general_voice = Channel(
        server_id=server.id,
        name="General",
        type=ChannelType.VOICE,
        position=0,
        top_position=1,
    )
    db.add_all([general_text, general_voice])
    await db.flush()

    return server


@router.get("/{server_id}", response_model=ServerDetailResponse)
async def get_server(
    server_id: int, user: CurrentUser, db: AsyncSession = Depends(get_db)
):
    """Get server details with channels user has access to."""
    result = await db.execute(
        select(Server)
        .where(Server.id == server_id)
        .options(selectinload(Server.channels))
    )
    server = result.scalar_one_or_none()
    if not server:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )
    
    # Check if user has access to this server
    if not check_server_access(user, server.min_server_level, server.min_internal_level):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="You do not have permission to access this server"
        )

    # Filter channels based on user's permission levels
    filtered_channels = [
        c for c in server.channels
        if check_channel_visibility(
            user,
            c.visibility_min_server_level,
            c.visibility_min_internal_level
        )
    ]

    return ServerDetailResponse(
        id=server.id,
        name=server.name,
        icon=server.icon,
        owner_id=server.owner_id,
        min_server_level=server.min_server_level,
        min_internal_level=server.min_internal_level,
        channels=[
            ChannelResponse(
                id=c.id,
                name=c.name,
                type=c.type.value,
                position=c.position,
                top_position=c.top_position,
                group_id=c.group_id,
                visibility_min_server_level=c.visibility_min_server_level,
                visibility_min_internal_level=c.visibility_min_internal_level,
                speak_min_server_level=c.speak_min_server_level,
                speak_min_internal_level=c.speak_min_internal_level,
            )
            for c in sorted(filtered_channels, key=lambda x: x.position)
        ],
    )


@router.patch("/{server_id}", response_model=ServerResponse)
@router.put(
    "/{server_id}", response_model=ServerResponse
)  # Keep for backward compatibility
async def update_server(
    server_id: int,
    payload: ServerUpdate,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Update a server (admin only)."""
    result = await db.execute(select(Server).where(Server.id == server_id))
    server = result.scalar_one_or_none()
    if not server:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    if payload.name is not None:
        server.name = payload.name
    if payload.icon is not None:
        server.icon = payload.icon
    if payload.min_server_level is not None:
        if not (1 <= payload.min_server_level <= 4):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="min_server_level must be between 1 and 4"
            )
        server.min_server_level = payload.min_server_level
    if payload.min_internal_level is not None:
        if not (1 <= payload.min_internal_level <= 2):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="min_internal_level must be 1 or 2"
            )
        server.min_internal_level = payload.min_internal_level

    await db.flush()
    return server


@router.delete("/{server_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_server(
    server_id: int, user: AdminUser, db: AsyncSession = Depends(get_db)
):
    """Delete a server (admin only)."""
    result = await db.execute(select(Server).where(Server.id == server_id))
    server = result.scalar_one_or_none()
    if not server:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    await db.delete(server)
    await db.commit()


# ============================================
# Unified Top-Level Reorder API
# ============================================


class ReorderItem(BaseModel):
    type: Literal["group", "channel"]
    id: int


class ReorderTopLevelRequest(BaseModel):
    """Request to reorder top-level items (groups and ungrouped channels)."""

    items: list[ReorderItem]


@router.post("/{server_id}/reorder")
async def reorder_top_level(
    server_id: int,
    payload: ReorderTopLevelRequest,
    user: AdminUser,
    db: AsyncSession = Depends(get_db),
):
    """Reorder top-level items: channel groups and ungrouped channels.

    This is the unified API for reordering the sidebar. Groups and ungrouped
    channels share a single position sequence (top_position for channels,
    position for groups).

    Example payload:
    {
        "items": [
            {"type": "group", "id": 1},
            {"type": "channel", "id": 5},
            {"type": "group", "id": 2}
        ]
    }
    """
    # Verify server exists
    server_result = await db.execute(select(Server).where(Server.id == server_id))
    if not server_result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    # Fetch all groups
    groups_result = await db.execute(
        select(ChannelGroup).where(ChannelGroup.server_id == server_id)
    )
    groups = {g.id: g for g in groups_result.scalars().all()}

    # Fetch all ungrouped channels
    channels_result = await db.execute(
        select(Channel).where(Channel.server_id == server_id, Channel.group_id == None)
    )
    ungrouped_channels = {c.id: c for c in channels_result.scalars().all()}

    # Validate all items
    seen_groups = set()
    seen_channels = set()

    for item in payload.items:
        if item.type == "group":
            if item.id not in groups:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Group {item.id} not found in server",
                )
            if item.id in seen_groups:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Duplicate group {item.id} in items",
                )
            seen_groups.add(item.id)
        elif item.type == "channel":
            if item.id not in ungrouped_channels:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Ungrouped channel {item.id} not found in server",
                )
            if item.id in seen_channels:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail=f"Duplicate channel {item.id} in items",
                )
            seen_channels.add(item.id)

    # Validate all items are provided
    if seen_groups != set(groups.keys()):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="All groups must be included in items",
        )
    if seen_channels != set(ungrouped_channels.keys()):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="All ungrouped channels must be included in items",
        )

    # Assign positions
    for idx, item in enumerate(payload.items):
        if item.type == "group":
            groups[item.id].position = idx
        else:
            ungrouped_channels[item.id].top_position = idx

    await db.flush()

    return {"success": True}


# Response models for all messages endpoint
class AttachmentResponse(BaseModel):
    id: int
    filename: str
    content_type: str
    size: int
    url: str

    class Config:
        from_attributes = True


class MentionResponse(BaseModel):
    """User mentioned in a message."""

    id: int
    username: str


class MessageInChannelResponse(UTCDateTimeModel):
    id: int
    channel_id: int
    user_id: int
    username: str
    avatar_url: str | None = None
    content: str
    created_at: datetime
    attachments: list[AttachmentResponse] = []
    is_deleted: bool = False
    edited_at: datetime | None = None
    mentions: list[MentionResponse] = []


class ChannelMessagesResponse(BaseModel):
    channel_id: int
    channel_name: str
    messages: list[MessageInChannelResponse]


@router.get("/{server_id}/all-messages", response_model=list[ChannelMessagesResponse])
async def get_all_server_messages(
    server_id: int,
    user: CurrentUser,
    db: AsyncSession = Depends(get_db),
    limit: int = Query(50, le=200, description="Max messages per channel"),
):
    """Get all messages from all text channels in a server.

    Returns a list of channel messages, where each item contains:
    - channel_id: The channel's ID
    - channel_name: The channel's name
    - messages: List of messages in that channel (up to limit per channel)
    """
    # Verify server exists
    server_result = await db.execute(select(Server).where(Server.id == server_id))
    server = server_result.scalar_one_or_none()
    if not server:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Server not found"
        )

    # Get all text channels in this server
    channels_result = await db.execute(
        select(Channel)
        .where(Channel.server_id == server_id, Channel.type == ChannelType.TEXT)
        .order_by(Channel.position)
    )
    channels = channels_result.scalars().all()

    # Get messages for each channel
    all_channel_messages = []
    for channel in channels:
        # Query messages for this channel
        messages_query = (
            select(Message)
            .where(
                Message.channel_id == channel.id,
                Message.is_deleted == False,
            )
            .options(selectinload(Message.attachments))
            .order_by(Message.id.desc())
            .limit(limit)
        )
        messages_result = await db.execute(messages_query)
        messages = messages_result.scalars().all()

        # Convert to response format
        message_responses = []
        for msg in reversed(messages):  # Reverse to get chronological order
            import re

            attachments = [
                AttachmentResponse(
                    id=att.id,
                    filename=att.filename,
                    content_type=att.content_type,
                    size=att.size,
                    url=f"/api/files/{att.id}",
                )
                for att in msg.attachments
            ]

            # Parse mentions from content
            mentions_data = [
                MentionResponse(id=0, username=username)
                for username in extract_mentioned_usernames(msg.content)
            ]

            message_responses.append(
                MessageInChannelResponse(
                    id=msg.id,
                    channel_id=msg.channel_id,
                    user_id=msg.user_id,
                    username=msg.username,
                    avatar_url=None,  # Can be populated from SSO if needed
                    content=msg.content,
                    created_at=msg.created_at,
                    attachments=attachments,
                    is_deleted=msg.is_deleted,
                    edited_at=msg.edited_at,
                    mentions=mentions_data,
                )
            )

        all_channel_messages.append(
            ChannelMessagesResponse(
                channel_id=channel.id,
                channel_name=channel.name,
                messages=message_responses,
            )
        )

    return all_channel_messages

File: backend-go/internal/handler/servers.go
package handler

import (
	"database/sql"
	"net/http"
	"strconv"
	"time"

	"github.com/labstack/echo/v4"

	"github.com/RMS-Server/rms-discord-go/internal/middleware"
	"github.com/RMS-Server/rms-discord-go/internal/permission"
	"github.com/RMS-Server/rms-discord-go/internal/sso"
)

// ServerHandler handles server CRUD and reorder endpoints.
type ServerHandler struct {
	db  *sql.DB
	sso *sso.Client
}

func NewServerHandler(db *sql.DB, sso *sso.Client) *ServerHandler {
	return &ServerHandler{db: db, sso: sso}
}

type serverCreateReq struct {
	Name string  `json:"name"`
	Icon *string `json:"icon"`
}

type serverUpdateReq struct {
	Name             *string `json:"name"`
	Icon             *string `json:"icon"`
	MinServerLevel   *int    `json:"min_server_level"`
	MinInternalLevel *int    `json:"min_internal_level"`
}

type serverResponse struct {
	ID               int64   `json:"id"`
	Name             string  `json:"name"`
	Icon             *string `json:"icon"`
	OwnerID          int64   `json:"owner_id"`
	MinServerLevel   int     `json:"min_server_level"`
	MinInternalLevel int     `json:"min_internal_level"`
}

type channelInServer struct {
	ID                         int64  `json:"id"`
	Name                       string `json:"name"`
	Type                       string `json:"type"`
	Position                   int    `json:"position"`
	TopPosition                int    `json:"top_position"`
	GroupID                    *int64 `json:"group_id"`
	VisibilityMinServerLevel   int    `json:"visibility_min_server_level"`
	VisibilityMinInternalLevel int    `json:"visibility_min_internal_level"`
	SpeakMinServerLevel        int    `json:"speak_min_server_level"`
	SpeakMinInternalLevel      int    `json:"speak_min_internal_level"`
}

type serverDetailResponse struct {
	serverResponse
	Channels []channelInServer `json:"channels"`
}

// ListServers returns servers filtered by user permission.
// GET /api/servers
func (h *ServerHandler) ListServers(c echo.Context) error {
	user := middleware.GetUser(c)

	rows, err := h.db.Query("SELECT id, name, icon, owner_id, min_server_level, min_internal_level FROM servers ORDER BY id")
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer rows.Close()

	var servers []serverResponse
	for rows.Next() {
		var s serverResponse
		if err := rows.Scan(&s.ID, &s.Name, &s.Icon, &s.OwnerID, &s.MinServerLevel, &s.MinInternalLevel); err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
		}
		if permission.CanAccessServer(user, s.MinServerLevel, s.MinInternalLevel) {
			servers = append(servers, s)
		}
	}
	if servers == nil {
		servers = []serverResponse{}
	}
	return c.JSON(http.StatusOK, servers)
}

// CreateServer creates a server with default channels.
// POST /api/servers
func (h *ServerHandler) CreateServer(c echo.Context) error {
	user := middleware.GetUser(c)
	var req serverCreateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}
	if req.Name == "" {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "name is required"})
	}

	tx, err := h.db.Begin()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer tx.Rollback()

	res, err := tx.Exec(
		"INSERT INTO servers (name, icon, owner_id, min_server_level, min_internal_level) VALUES (?, ?, ?, 1, 1)",
		req.Name, req.Icon, user.ID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	serverID, _ := res.LastInsertId()

	// Create default text and voice channels
	_, err = tx.Exec(
		"INSERT INTO channels (server_id, name, type, position, top_position, visibility_min_server_level, visibility_min_internal_level, speak_min_server_level, speak_min_internal_level) VALUES (?, 'general', 'TEXT', 0, 0, 1, 1, 1, 1)",
		serverID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	_, err = tx.Exec(
		"INSERT INTO channels (server_id, name, type, position, top_position, visibility_min_server_level, visibility_min_internal_level, speak_min_server_level, speak_min_internal_level) VALUES (?, 'General', 'VOICE', 0, 1, 1, 1, 1, 1)",
		serverID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	if err := tx.Commit(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusCreated, serverResponse{
		ID: serverID, Name: req.Name, Icon: req.Icon,
		OwnerID: int64(user.ID), MinServerLevel: 1, MinInternalLevel: 1,
	})
}

// GetServer returns server detail with filtered channels.
// GET /api/servers/:id
func (h *ServerHandler) GetServer(c echo.Context) error {
	user := middleware.GetUser(c)
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	var s serverResponse
	err = h.db.QueryRow("SELECT id, name, icon, owner_id, min_server_level, min_internal_level FROM servers WHERE id = ?", id).
		Scan(&s.ID, &s.Name, &s.Icon, &s.OwnerID, &s.MinServerLevel, &s.MinInternalLevel)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	if !permission.CanAccessServer(user, s.MinServerLevel, s.MinInternalLevel) {
		return c.JSON(http.StatusForbidden, map[string]string{"error": "you do not have permission to access this server"})
	}

	rows, err := h.db.Query(
		"SELECT id, name, type, position, top_position, group_id, visibility_min_server_level, visibility_min_internal_level, speak_min_server_level, speak_min_internal_level FROM channels WHERE server_id = ? ORDER BY top_position, position",
		id,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer rows.Close()

	var channels []channelInServer
	for rows.Next() {
		var ch channelInServer
		if err := rows.Scan(&ch.ID, &ch.Name, &ch.Type, &ch.Position, &ch.TopPosition, &ch.GroupID,
			&ch.VisibilityMinServerLevel, &ch.VisibilityMinInternalLevel,
			&ch.SpeakMinServerLevel, &ch.SpeakMinInternalLevel); err != nil {
			return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
		}
		if permission.CanSeeChannel(user, ch.VisibilityMinServerLevel, ch.VisibilityMinInternalLevel) {
			channels = append(channels, ch)
		}
	}
	if channels == nil {
		channels = []channelInServer{}
	}

	return c.JSON(http.StatusOK, serverDetailResponse{serverResponse: s, Channels: channels})
}

// UpdateServer updates server properties.
// PATCH /api/servers/:id  (also PUT for backward compat)
func (h *ServerHandler) UpdateServer(c echo.Context) error {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	var s serverResponse
	err = h.db.QueryRow("SELECT id, name, icon, owner_id, min_server_level, min_internal_level FROM servers WHERE id = ?", id).
		Scan(&s.ID, &s.Name, &s.Icon, &s.OwnerID, &s.MinServerLevel, &s.MinInternalLevel)
	if err == sql.ErrNoRows {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	var req serverUpdateReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}

	if req.Name != nil {
		s.Name = *req.Name
	}
	if req.Icon != nil {
		s.Icon = req.Icon
	}
	if req.MinServerLevel != nil {
		if *req.MinServerLevel < 1 || *req.MinServerLevel > 4 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_server_level must be between 1 and 4"})
		}
		s.MinServerLevel = *req.MinServerLevel
	}
	if req.MinInternalLevel != nil {
		if *req.MinInternalLevel < 1 || *req.MinInternalLevel > 2 {
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "min_internal_level must be 1 or 2"})
		}
		s.MinInternalLevel = *req.MinInternalLevel
	}

	_, err = h.db.Exec(
		"UPDATE servers SET name = ?, icon = ?, min_server_level = ?, min_internal_level = ? WHERE id = ?",
		s.Name, s.Icon, s.MinServerLevel, s.MinInternalLevel, id,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}

	return c.JSON(http.StatusOK, s)
}

// DeleteServer deletes a server.
// DELETE /api/servers/:id
func (h *ServerHandler) DeleteServer(c echo.Context) error {
	id, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	res, err := h.db.Exec("DELETE FROM servers WHERE id = ?", id)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	n, _ := res.RowsAffected()
	if n == 0 {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}
	return c.NoContent(http.StatusNoContent)
}

type reorderItem struct {
	Type string `json:"type"` // "group" or "channel"
	ID   int64  `json:"id"`
}

type reorderTopLevelReq struct {
	Items []reorderItem `json:"items"`
}

// ReorderTopLevel reorders groups and ungrouped channels.
// POST /api/servers/:id/reorder
func (h *ServerHandler) ReorderTopLevel(c echo.Context) error {
	serverID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	var req reorderTopLevelReq
	if err := c.Bind(&req); err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid request body"})
	}

	// Verify server exists
	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM servers WHERE id = ?", serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}

	// Fetch existing groups
	groupRows, err := h.db.Query("SELECT id FROM channel_groups WHERE server_id = ?", serverID)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	groupIDs := map[int64]bool{}
	for groupRows.Next() {
		var gid int64
		groupRows.Scan(&gid)
		groupIDs[gid] = true
	}
	groupRows.Close()

	// Fetch ungrouped channels
	chRows, err := h.db.Query("SELECT id FROM channels WHERE server_id = ? AND group_id IS NULL", serverID)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	chIDs := map[int64]bool{}
	for chRows.Next() {
		var cid int64
		chRows.Scan(&cid)
		chIDs[cid] = true
	}
	chRows.Close()

	// Validate items
	seenGroups := map[int64]bool{}
	seenChannels := map[int64]bool{}
	for _, item := range req.Items {
		switch item.Type {
		case "group":
			if !groupIDs[item.ID] {
				return c.JSON(http.StatusBadRequest, map[string]string{"error": "group not found in server"})
			}
			if seenGroups[item.ID] {
				return c.JSON(http.StatusBadRequest, map[string]string{"error": "duplicate group in items"})
			}
			seenGroups[item.ID] = true
		case "channel":
			if !chIDs[item.ID] {
				return c.JSON(http.StatusBadRequest, map[string]string{"error": "ungrouped channel not found in server"})
			}
			if seenChannels[item.ID] {
				return c.JSON(http.StatusBadRequest, map[string]string{"error": "duplicate channel in items"})
			}
			seenChannels[item.ID] = true
		default:
			return c.JSON(http.StatusBadRequest, map[string]string{"error": "item type must be 'group' or 'channel'"})
		}
	}

	if len(seenGroups) != len(groupIDs) {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "all groups must be included in items"})
	}
	if len(seenChannels) != len(chIDs) {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "all ungrouped channels must be included in items"})
	}

	tx, err := h.db.Begin()
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer tx.Rollback()

	for idx, item := range req.Items {
		if item.Type == "group" {
			tx.Exec("UPDATE channel_groups SET position = ? WHERE id = ?", idx, item.ID)
		} else {
			tx.Exec("UPDATE channels SET top_position = ? WHERE id = ?", idx, item.ID)
		}
	}

	if err := tx.Commit(); err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	return c.JSON(http.StatusOK, map[string]bool{"success": true})
}

// GetAllMessages returns messages from all text channels in a server.
// GET /api/servers/:id/all-messages
func (h *ServerHandler) GetAllMessages(c echo.Context) error {
	user := middleware.GetUser(c)
	serverID, err := strconv.ParseInt(c.Param("id"), 10, 64)
	if err != nil {
		return c.JSON(http.StatusBadRequest, map[string]string{"error": "invalid server id"})
	}

	limitStr := c.QueryParam("limit")
	limit := 50
	if limitStr != "" {
		if v, err := strconv.Atoi(limitStr); err == nil && v > 0 && v <= 200 {
			limit = v
		}
	}

	// Verify server
	var exists int
	if err := h.db.QueryRow("SELECT 1 FROM servers WHERE id = ?", serverID).Scan(&exists); err != nil {
		return c.JSON(http.StatusNotFound, map[string]string{"error": "server not found"})
	}

	// Get text channels
	chRows, err := h.db.Query(
		"SELECT id, name, visibility_min_server_level, visibility_min_internal_level FROM channels WHERE server_id = ? AND type = 'TEXT' ORDER BY position",
		serverID,
	)
	if err != nil {
		return c.JSON(http.StatusInternalServerError, map[string]string{"error": err.Error()})
	}
	defer chRows.Close()

	type channelInfo struct {
		ID                         int64
		Name                       string
		VisibilityMinServerLevel   int
		VisibilityMinInternalLevel int
	}
	var textChannels []channelInfo
	for chRows.Next() {
		var ch channelInfo
		chRows.Scan(&ch.ID, &ch.Name, &ch.VisibilityMinServerLevel, &ch.VisibilityMinInternalLevel)
		if permission.CanSeeChannel(user, ch.VisibilityMinServerLevel, ch.VisibilityMinInternalLevel) {
			textChannels = append(textChannels, ch)
		}
	}

	type msgResp struct {
		ID        int64   `json:"id"`
		ChannelID int64   `json:"channel_id"`
		UserID    int64   `json:"user_id"`
		Username  string  `json:"username"`
		AvatarURL *string `json:"avatar_url"`
		Content   string  `json:"content"`
		CreatedAt string  `json:"created_at"`
		IsDeleted bool    `json:"is_deleted"`
		EditedAt  *string `json:"edited_at"`
	}

	type channelMsgsResp struct {
		ChannelID   int64     `json:"channel_id"`
		ChannelName string    `json:"channel_name"`
		Messages    []msgResp `json:"messages"`
	}

	result := make([]channelMsgsResp, 0, len(textChannels))
	for _, ch := range textChannels {
		rows, err := h.db.Query(
			"SELECT id, channel_id, user_id, username, content, created_at, is_deleted, edited_at FROM messages WHERE channel_id = ? AND is_deleted = FALSE ORDER BY id DESC LIMIT ?",
			ch.ID, limit,
		)
		if err != nil {
			continue
		}

		var msgs []msgResp
		for rows.Next() {
			var m msgResp
			var createdAt time.Time
			var editedAt sql.NullTime
			rows.Scan(&m.ID, &m.ChannelID, &m.UserID, &m.Username, &m.Content, &createdAt, &m.IsDeleted, &editedAt)
			m.CreatedAt = createdAt.UTC().Format("2006-01-02T15:04:05Z")
			if editedAt.Valid {
				s := editedAt.Time.UTC().Format("2006-01-02T15:04:05Z")
				m.EditedAt = &s
			}
			msgs = append(msgs, m)
		}
		rows.Close()

		// Reverse for chronological order
		for i, j := 0, len(msgs)-1; i < j; i, j = i+1, j-1 {
			msgs[i], msgs[j] = msgs[j], msgs[i]
		}
		if msgs == nil {
			msgs = []msgResp{}
		}

		result = append(result, channelMsgsResp{ChannelID: ch.ID, ChannelName: ch.Name, Messages: msgs})
	}

	return c.JSON(http.StatusOK, result)
}
